### PR TITLE
[ECO] Fast Indexer

### PR DESCRIPTION
**Pull request by CapCap**
### Description

Separate threads for fetching data, and separate threads for pushing it to DB in parallel

increases (locally) performance of (old) indexer from 1.4k TPS, to 2.7k TPS (locally, running against sherrynet)

Pushing from local to RDS performance went from 20TPS to ~150 TPS with bulk fetch, and to 600 TPS with batch write.

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3372)

### CODE REVIEW

File: ecosystem/indexer/src/indexer/fetcher.rs [hunk 1]
```diff
@@ -3,97 +3,176 @@
...
+
+    panic!(
+        "Could not fetch {} transactions starting at {}!",
+        TRANSACTION_FETCH_BATCH_SIZE, starting_version
+    );
+}
+
+#[derive(Debug)]
+pub struct TransactionFetcher {
+    starting_version: u64,
```
Comment (bowenyang007) [inline ecosystem/indexer/src/indexer/fetcher.rs:145]: Given we don't update the starting_version, why not just pass it as a param in start()?

**Review by bowenyang007 — APPROVED** [review]
Generally makes sense, I have some questions about implementation details but given the short lifespan of this indexer let's just go for it. It can't be worse...right?

File: ecosystem/indexer/src/indexer/fetcher.rs [hunk 2]
```diff
@@ -3,97 +3,176 @@
...
+            let num_batches = std::cmp::min(
+                (num_missing as f64 / TRANSACTION_FETCH_BATCH_SIZE as f64).ceil() as u64,
+                MAX_THREADS as u64,
+            ) as usize;
+            let mut futures = vec![];
+            for i in 0..num_batches {
+                futures.push(fetch_nexts(
+                    self.client.clone(),
+                    self.current_version + (i as u64 * TRANSACTION_FETCH_BATCH_SIZE as u64),
+                ));
```
[the remark this answers is not in the document]
Comment (CapCap) [reply ecosystem/indexer/src/indexer/fetcher.rs:76]: oh snap I think this is a bug

[the remark this answers is not in the document]
Comment (CapCap) [reply ecosystem/indexer/src/indexer/fetcher.rs:76]: updated, good catch!

File: ecosystem/indexer/src/indexer/fetcher.rs [hunk 3]
```diff
@@ -3,97 +3,176 @@
...
+                self.set_highest_known_version().await;
+            }
+
+            let num_missing = self.highest_known_version - self.current_version;
+            let num_batches = std::cmp::min(
+                (num_missing as f64 / TRANSACTION_FETCH_BATCH_SIZE as f64).ceil() as u64,
+                MAX_THREADS as u64,
+            ) as usize;
+            let mut futures = vec![];
+            for i in 0..num_batches {
```
[the remark this answers is not in the document]
Comment (CapCap) [reply ecosystem/indexer/src/indexer/fetcher.rs:72]: yeah- this is something I hope you can fix in your batch updated PR ;-)

**Pull request merged**